### ISSUE TITLE
Changing ports (Raspberry Pi)..?

### ISSUE THREAD

**Issue by 4C3T** [sheharyaar/ptokax#23]
Hi. Just installed this on my Pi Zero. Initially had a few errors while compiling, but re-ran it a few times and some of the files created (executables) and after a while the problems went away.

Now, however, I'm not entirely sure but either PtokaX is overwriting the cfg.pxt (Seems to very much so be the case when running the ./PtokaX), and if the service isn't stopped properly before changing it (file lock perhaps?)

However, initially I edited the file and added port 7878, but I was unable to enter this. I found that the cfg-file had been replaced several times, other and most of the times not.
When I then shrugged my shoulders and tried the default ports 1209 and 411 it just worked. Also saw something in a script file that the service had been allowed for ports below some 1xxx (thousand).

I'm no Linux expert, but also not born yesterday, still I'm at a loss here. Shouldn't I be able to use whichever port I want?
Should also point out that I've never used the Linux version before so the environment is still a bit alien compared to Windows, and I'm no real Linux enthusiast either, so my knowledge is somewhat limited, and usually Google helps compiling and fixing (some) errors or getting dumber by the line (of RTFM).

**Comment by sheharyaar** [comment]
Did you try to set the port in this file ?
https://github.com/sheharyaar/ptokax/blob/main/PtokaX/cfg/Settings.pxt

Also, please don't forget to uncomment the line

Do reach out again if this does not fix it.

**Comment by sheharyaar** [comment]
Changing this won't have any effect, this is just example config. You need to change the ports in the file mentioned in my prev comment https://github.com/sheharyaar/ptokax/blob/main/PtokaX/cfg.example/Settings.pxt

**Comment by 4C3T** [comment]
> Did you try to set the port in this file ? https://github.com/sheharyaar/ptokax/blob/main/PtokaX/cfg/Settings.pxt
> 
> Also, please don't forget to uncomment the line
> 
> Do reach out again if this does not fix it.

You're referring to these lines:
`# TCP ports. Minimal length 1. Maximal length 64

#TCPPorts	=	1209;411`




> Changing this won't have any effect, this is just example config. You need to change the ports in the file mentioned in my prev comment https://github.com/sheharyaar/ptokax/blob/main/PtokaX/cfg.example/Settings.pxt

Yes, of course the path's not the correct one. I found the correct one to change, as I more or less mentioned. I was able to use the original ports, but was unable to change them (A.K.A. didn't start at all, or did start but unable to communicate - I don't know).

**Comment by sheharyaar** [comment]
Ok I will have a look into this, for now I don't have access to a raspberry pi.

**Comment by 4C3T** [comment]
It looks to me that the service script ptokax-setup.sh (below) which I don't exactly remember, but I think I downloaded it from a site when I searched for how to install it.

There's a line in that which says "**_# Setup priveleges in order to allow ptokax to run on port 411 (default port)_**" but then there's nothing to indicate an actual port being set, just some LED's indications (which are somewhat undocumented):

> #!/bin/bash
> 
> set -eou pipefail
> 
> RED=$(tput setaf 1)
> GREEN=$(tput setaf 2)
> YELLOW=$(tput setaf 3)
> BLUE=$(tput setaf 4)
> WHITE=$(tput setaf 7)
> ORANGE=$(tput setaf 9)
> 
> echo -e "${GREEN}[+] ${BLUE}Enabling and Starting DHCP service${WHITE}"
> sudo service dhcpcd start
> sudo systemctl enable dhcpcd 2>/dev/null
> 
> # Adding the static ip config only if an entry doesn't already exist
> STATIC_IP=$(grep -q '#Static IP for PtokaX' /etc/dhcpcd.conf && echo true || echo false)
> if [ "$STATIC_IP" == "false" ]; then
> 	echo -e "${GREEN}[+] ${BLUE}Making RPi's IP address static${WHITE}"
> 
> 	RASPI_IP=$(ip addr | grep inet | grep eth0 | cut -d' ' -f6)
> 	GATEWAY_IP=$(ip route | grep default | cut -d' ' -f3)
> 	DNS_ADDRESS=$(grep nameserver -m 1 /etc/resolv.conf | cut -d' ' -f2)
> 
> 	cat <<- EOF >> /etc/dhcpcd.conf
> 		#Static IP for PtokaX
> 		interface eth0
> 		static ip_address=${RASPI_IP}
> 		static routers=${GATEWAY_IP}
> 		static domain_name_servers=${DNS_ADDRESS}
> 	EOF
> else
> 	echo -e "${YELLOW}[-] ${BLUE}RPi's IP address is already static${WHITE}"
> fi
> 
> echo -e "${GREEN}[+] ${BLUE}Downloading PtokaX - start, stop & setup scripts${WHITE}"
> for action in "start" "stop" "setup"; do
> 	curl -s https://raw.githubusercontent.com/sheharyaar/ptokax/main/ptokax-${action}.sh -L -o ~/ptokax-${action}.sh
> 	chmod +x ~/ptokax-${action}.sh
> done
> 
> echo -e "${GREEN}[+] ${BLUE}Installing Prerequisites${WHITE}"
> # Curl for downloading source code
> # mysql - required for scripts
> # LUA 5.2.2 - Latest lua not installed as scripts are in lua version 5.2.2
> sudo apt install -y curl liblua5.2-dev make g++ zlib1g-dev libtinyxml-dev default-libmysqlclient-dev lua-sql-mysql
> 
> echo -e "${GREEN}[+] ${BLUE}Downloading PtokaX${WHITE}"
> # Download PtokaX source code
> curl -L -s https://github.com/sheharyaar/ptokax/releases/download/latest/ptokax-0.5.2.2-src.tgz -o ~/ptokax-0.5.2.2-src.tgz
> # Extract the archive
> tar -xf ~/ptokax-0.5.2.2-src.tgz
> echo -e "${GREEN}[+] ${BLUE}Installing PtokaX${WHITE}"
> # Make the program
> cd PtokaX/ || (echo "cd to PtokaX failed" && exit)
> make -f makefile-mysql lua52
> # Setup privileges in order to allow ptokax to run on port 411 (default port)
> sudo apt install -y libcap2-bin
> sudo make install
> 
> echo -e "${GREEN}[+] ${BLUE}Setting up PtokaX${WHITE}"
> ./PtokaX -m
> 
> if [ ! -d ~/ptokax-scripts ]; then
> 	echo -e "${GREEN}[+] ${BLUE}Downloading Hit Hi Fit Hai scripts${WHITE}"
> 	git clone https://github.com/sheharyaar/ptokax-scripts ~/ptokax-scripts
> else
> 	echo -e "${YELLOW}[-] ${BLUE}Hit Hi Fit Hai scripts already exist${WHITE}"
> fi
> cp ~/ptokax-scripts/* ~/PtokaX/scripts/ -rf
> 
> echo -e "${GREEN}[+] ${BLUE}Enabling and starting PtokaX service${WHITE}"
> sudo systemctl enable ptokax.service
> sudo service ptokax.service start
> 
> echo -e "${GREEN}[+] ${BLUE}Enabling and starting PtokaX-DHCP service${WHITE}"
> sudo systemctl enable ptokax-dhcp.service
> sudo service ptokax-dhcp.service start
> 
> echo -e "${RED}[*] ${BLUE}Run PtokaX server using ${ORANGE}sudo ~/ptokax-start.sh${WHITE}"
> 
> echo -e "${GREEN}[+] ${BLUE}Rebooting Pi in 10 seconds ...${WHITE}"
> sleep 10; sudo reboot

Ran this once and it installed most of it, but didn't install the service (no ptokax-dhcp.service file in /etc/systemd/system), but the files ptokax-start.sh and ptokax-stop.sh works. I had some errors so I ran a couple of other (more specific) scripts within the folders while compiling the thing.

This hub is very over-complicated.
I'm also running luadch and it's nowhere near as complex. Just make sure the pre-reqs are installed, then just run it and you can configure it however you want and it just runs.

This is totally different. And I don't understand why the DHCP need to be setup either since there's a setting for it (in Settings.pxt) where you can specify which interfaces or IP to be used.
Also, that file is very compacted so it's very hard on the eye to read & make changes to.

I'm just running it since I need a Neomodus-compatible  hub with familiar scripts to run off my Pi Zero. The LEDs function is also very interesting.

**Comment by 4C3T** [comment]
Sorry for the formatting

**Comment by sheharyaar** [comment]
This is still a :wip `ptokax-setup.sh`. We use it to setup ptokax on our system. I will move it to another folder to let others know that its not related to ptokax

The part `Setup priveleges in order to allow ptokax to run on port 411 (default port)` is used to add capability to ptokax to be able to create ports under 1024 (this is generally not allowed in linux unless you use capability or root privileges). It is not related to port setting

**Comment by sheharyaar** [comment]
Also, Ptokax is maintained at http://www.ptokax.org/downloads.html. This repo just has scripts to setup Ptokax in my university, so it has the modifications we needed to set it up. We are not the owners of the project.

**Comment by sheharyaar** [comment]
You can follow the original post here to setup ptokax on raspbian : http://wiki.ptokax.org/doku.php?id=guides:debian_bugbuntu

**Comment by 4C3T** [comment]
Ok. So, a small update.
I'm not used to LUA since it's pretty much retarded to begin with.
I've managed to change the ports now.
The problem is that... Well, I don't know where to begin.

Part of the solution though is to uncomment the settings.

I first thought that they were supposed to look like that since the comments had "# then the text", while the setting had "#SettingGoesHere" with no space.

Here's the kicker - Some settings that I changed, despite the line having a hashtag, did make a difference, to some stuff.
I'm also experiencing, still, what appears to be some either some sort of weird file lock, or the shutdown command makes it write over the Settings.pxt and also Script.pxt.

When I edit the files locally or remotely I can verify immediately afterwards. 
I just tried removing the rights (chmod) for writing and one setting that I tested to change was the default bot which is named "PtokaX" which I set to "Memorial Bot" - Totally refused to change it, but it does read the description for it, oddly enough.

This makes Zero sense to me.

I did have some trouble about a month ago with luadch, but I was unable to just copy the files like you normally do since all files are read into RAM and no file locks in place. If any file's been changed, you just issue a !reload or !restart and it'll reload those.
However, I was unable to actually write over files. I just didn't get any errors (In mc), so when I did a refresh the old files were still there.
Shutting down luadch solved that problem and I only had that problem twice.

I just checked top (command), and I can't see PtokaX running after I've run that stop-script.
Also, file rights are somewhat of a nightmare. I explicitly set read-only for Settings.pxt, yet I was able to edit & save the data (Checked by opening again and added/changed text was there!)...

This whole thing is just weird!
Should also mention that I'm running as "sudo su" which means that I have full admin access...

When running the startup-script I see this:
"root@movielovershub:~# ./ptokax-start.sh
Starting PtokaX DC Hub 0.5.2.2 as daemon using /root/PtokaX as config directory.
[+] Successfully started PtokaX server!"

The path is correct too, and I can't think of any reason why it reads some (and leaves them untouched) lines, and defaults others in the Settings.pxt (and also Scripts.pxt)

The latter is also weird.
I had lines at first with `#scriptname.lua <tab>=<tab>1`
For some reason, the hashtag was removed, and the value set to 0
When changed manually to one, it goes back to 0 after a restart.
Scripts error logs says something about errors bla bla bla, but regardless, I don't see why it disables them in the settings file...
